src/Application/GameStarter.cpp

```diff
@@ -183,44 +197,86 @@ GameStarter::~GameStarter() {
-                }
-            }
+    std::vector<std::string> candidates;
+    if (!options->dataPath.empty()) {
+        candidates.push_back(options->dataPath);
+    } else {
+        candidates = resolveMm7Paths(environment);
+    }
+    assert(!candidates.empty());
+
+    for (int i = 0; i < candidates.size(); i++) {
+        std::string missingFile;
+        if (!std::filesystem::exists(candidates[i])) {
+            logger->info("Data path #{} ('{}') doesn't exist.", i + 1, candidates[i]);
+        } else if (!validateMm7Path(candidates[i], &missingFile)) {
+            logger->info("Data path #{} ('{}') is missing file '{}'.", i + 1, candidates[i], missingFile);
+        } else {
+            logger->info("Data path #{} ('{}') is OK!", i + 1, candidates[i]);
+            options->dataPath = candidates[i];
+            break;
         }
-
-        if (options->dataPath.empty()) // Only one candidate, or no valid candidates? Use the last one & re-check it later.
-            options->dataPath = candidates.back();
     }
 
-    assert(!options->dataPath.empty());
+    // Just use the last data path if all paths are invalid. We'll throw later.
+    if (options->dataPath.empty())
+        options->dataPath = candidates.back();
     logger->info("Using data path '{}'.", options->dataPath);
 
-    if (options->useConfig && options->configPath.empty())
-        options->configPath = options->dataPath + "/openenroth.ini";
+    if (options->userPath.empty())
+        options->userPath = resolveMm7UserPath(environment);
+    logger->info("Using user path '{}'.", options->userPath);
+}
+
+void GameStarter::failOnInvalidPath(std::string_view dataPath, Platform *platform) {
+    std::string missingFile;
+    if (validateMm7Path(dataPath, &missingFile))
+        return;
+
+    std::string message = fmt::format(
+        "Required file '{}' not found.\n"
+        "You should acquire licensed copy of M&M VII and copy its resources to \n{}\n\n"
+        "Additionally you should also copy the content from\n"
+        "resources directory from our repository there as well.",
+        missingFile,
+        dataPath
+    );
+    logger->critical("{}", message);
+    platform->showMessageBox("CRITICAL ERROR: missing resources", message);
+    throw Exception("Data folder '{}' validation failed", dataPath);
+}
+
+void GameStarter::migrateUserData() {
+    logger->info("Migrating user data from '{}' to '{}'...", dfs->displayPath(""), ufs->displayPath(""));
+
+    if (ufs->exists("saves") && !ufs->ls("saves").empty()) {
+        logger->info("    Target saves directory is not empty, skipping saves migration.");
+    } else if (!dfs->exists("saves")) {
+        logger->info("    No save files to migrate.");
+    } else {
+        for (const DirectoryEntry &entry : dfs->ls("saves")) {
+            std::string path = fmt::format("saves/{}", entry.name);
+            ufs->write(path, dfs->read(path));
+            logger->info("    Copied '{}'.", entry.name);
```

Comment:
This can trip up if like me you have subfolders in the saves dir. Atypical use though so maybe just check entry is a file before trying to copy?